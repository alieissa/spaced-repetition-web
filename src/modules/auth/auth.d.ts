   type UserLogin = {
    email: string,
    password: string,
  }
   type UserToken = {
    token: string
  }
   type RememberMeUserForm = UserLogin & {
    rememberMe: boolean
  }
   type ForgotPasswordForm = {
    email: string
  }
   type ResetPasswordForm = {
    password: string
    confirmedPassword: string
   }
