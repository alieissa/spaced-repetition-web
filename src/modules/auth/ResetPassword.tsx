import { FormikState, useFormik } from 'formik'
import { ChangeEvent, useEffect } from 'react'
import { Container, Form, Message } from 'semantic-ui-react'
import { SPButton, SPHeader, SPInput, SPText } from 'src/components'
import { styles } from 'src/styles'
import { async } from 'src/utils'
import * as Yup from 'yup'
import { useResetPassword } from './auth.hooks'
import { NAuth } from './auth.types'

const ResetPasswordValidatonSchema = Yup.object().shape({
  password: Yup.string()
    .required('Required')
    .min(5, 'Password must be at least 5 characters.'),
  confirmededPassword: Yup.string().oneOf(
    [Yup.ref('password')],
    'Passwords must match',
  ),
})

type Props = {
  form: FormikState<NAuth.ResetPasswordForm>
  onChangePassword: (e: ChangeEvent<HTMLInputElement>) => void
  onChangeConfirmedPassword: (e: ChangeEvent<HTMLInputElement>) => void
  onResetPassword: VoidFunction
}
function ResetPasswordForm(props: Props) {
  const passwordError =
    props.form.touched.password && !!props.form.errors.password
  const confirmedPasswordError =
    props.form.touched.confirmedPassword &&
    !!props.form.errors.confirmedPassword
  return (
    <Form
      data-testid="reset-password-form"
      className="flex-column bordered p-1r"
      onSubmit={props.onResetPassword}
    >
      <Form.Field error={passwordError || confirmedPasswordError}>
        <label htmlFor="rest-password-password-input">Password</label>
        <SPInput
          id="rest-password-password-input"
          data-testid="reset-password-form-password-input"
          title="password"
          disabled={props.form.isSubmitting}
          placeholder="Password"
          type="password"
          value={props.form.values.password}
          onChange={props.onChangePassword}
        />
        {passwordError && (
          <SPText
            data-testid="reset-password-form-password-error"
            style={styles['mt-0.5rem']}
          >
            {props.form.errors.password}
          </SPText>
        )}
      </Form.Field>
      <Form.Field error={confirmedPasswordError}>
        <label htmlFor="reset-password-confirm-password-input">
          Confirm password
        </label>
        <SPInput
          id="reset-password-confirm-password-input"
          data-testid="reset-password-form-confirmed-password-input"
          title="confirm password"
          disabled={props.form.isSubmitting}
          placeholder="Confirm password"
          type="password"
          value={props.form.values.confirmedPassword}
          onChange={props.onChangeConfirmedPassword}
        />
        {confirmedPasswordError && (
          <SPText
            data-testid="reset-password-form-confirmed-password-error"
            style={styles['mt-0.5rem']}
          >
            {props.form.errors.confirmedPassword}
          </SPText>
        )}
      </Form.Field>
      <SPButton
        data-testid="reset-password-form-btn"
        className="align-self-end"
        type="submit"
        disabled={passwordError || confirmedPasswordError}
      >
        Submit
      </SPButton>
    </Form>
  )
}

export default function ResetPassword() {
  const [resetPasswordStatus, resetPassword] = useResetPassword()

  const form = useFormik({
    initialValues: {
      password: '',
      confirmedPassword: '',
    },
    validationSchema: ResetPasswordValidatonSchema,
    onSubmit: (values: NAuth.ResetPasswordForm) =>
      resetPassword({ password: values.password }),
  })

  useEffect(() => {
    form.setSubmitting(resetPasswordStatus.type === 'Loading')
  }, [resetPasswordStatus])

  const handleChangePassword = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('password', e.target.value)
  }

  const handleChangeConfirmedPassword = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('confirmedPassword', e.target.value)
  }

  const commonProps = {
    form: form,
    onChangePassword: handleChangePassword,
    onChangeConfirmedPassword: handleChangeConfirmedPassword,
    onResetPassword: form.submitForm,
  }
  return (
    <Container>
      <div className="justify-space-between bordered p-1r mb-1r">
        <SPHeader as="h1">Spaced Reps</SPHeader>
      </div>

      {async.match(resetPasswordStatus)({
        Untriggered: () => <ResetPasswordForm {...commonProps} />,
        Loading: () => <ResetPasswordForm {...commonProps} />,
        Failure: () => (
          <>
            <div data-testid="reset-password-form-error">
              <Message negative>Unable to reset password</Message>
            </div>
            <ResetPasswordForm {...commonProps} />
          </>
        ),
        Success: () => (
          <div data-testid="reset-password-form-success">
            <Message positive>Your password has been reset</Message>
          </div>
        ),
      })}
    </Container>
  )
}
