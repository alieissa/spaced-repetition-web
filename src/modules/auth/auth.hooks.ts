/** @format */

import { Left, Right } from 'src/utils/either'
type UseRequestParams = {
  url: string
  method: 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE'
}
/**
 * Returns a fetch that has the access token included in the header.
 * Auth0 only gives access to access token from a hook, so forced to
 * create a hook around it. There is the option to save the access
 * token in localStorage but that is not recommended and is also
 * difficult because there is no easy way to know the key under which
 * the access token is saved
 */
export function useAuthRequest<D = {}>({ url, method }: UseRequestParams) {
  const token = localStorage.getItem('token')
  const authFetch = (data?: D) => {
    const apiUrl = `${process.env.REACT_APP_API_ENDPOINT}/${url}`
    const headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }
    const init = data
      ? { method, headers, body: JSON.stringify(data) }
      : { method, headers }

    return fetch(apiUrl, init)
      .then((r) => r.json())
      .then((data) => Right<D>(data))
      .catch((error) => Left(error))
  }
  return authFetch
}
