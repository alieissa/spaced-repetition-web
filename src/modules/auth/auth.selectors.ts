import { RootState } from "src/types";

export const status = (state: RootState) => {
  return state.auth.status
}

export const logoutStatus = (state: RootState) => {
  return state.auth.logoutStatus
}