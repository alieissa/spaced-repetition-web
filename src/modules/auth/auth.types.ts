import { Async } from 'src/utils/async'
export namespace NAuth {
  export type State = {
    status: Async<null, Error['message'], UserToken>
    logoutStatus: Async<null, Error['message'], string>
  }

  export type User = {
    email: string,
    password: string,
  }
  export type UserToken = {
    token: string
  }
  export type UserForm = User & {
    rememberMe: boolean
  }
}