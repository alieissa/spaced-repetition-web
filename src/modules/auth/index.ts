export { default as ForgotPassword } from './ForgotPassword';
export { default as Login } from './Login';
export { default as ResetPassword } from './ResetPassword';
export { default as authReducer } from './auth.reducer';
export type { NAuth } from './auth.types';

