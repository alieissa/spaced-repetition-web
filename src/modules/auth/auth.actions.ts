/** @format */

import { Either } from 'src/utils/either'
import { NAuth } from './auth.types'

export type Login = {
  type: 'Login'
}

export type LoggedIn = {
  type: 'LoggedIn'
  rememberMe: boolean,
  result: Either<Error['message'], NAuth.UserToken>
}

export type ResetLogin = {
  type: 'ResetLogin'
}

export type Logout = {
  type: 'Logout'
}

export type LoggedOut = {
  type: 'LoggedOut'
  result: Either<Error['message'], string>
}

export type LoginAction = Login | LoggedIn | ResetLogin | Logout | LoggedOut
