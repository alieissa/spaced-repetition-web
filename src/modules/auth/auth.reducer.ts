/** @format */

import produce from 'immer'
import { either } from 'src/utils'
import { Failure, Loading, Success, Untriggered } from 'src/utils/async'
import type { LoginAction } from './auth.actions'
import { NAuth } from './auth.types'

const initialState: NAuth.State = {
  status: Untriggered(),
  logoutStatus: Untriggered(),
  notifyForgotPasswordStatus: Untriggered(),
}

export default produce((draft: NAuth.State, action: LoginAction) => {
  switch (action.type) {
    case 'Login': {
      draft.status = Loading(null)
      draft.logoutStatus = Untriggered()
      return
    }
    case 'LoggedIn': {
      either.match(action.result)({
        Left: ({ value }) => {
          draft.status = Failure(value)
        },
        Right: ({ value }) => {
          const storage = action.rememberMe ? localStorage : sessionStorage
          storage.setItem('token', value.token)
          draft.status = Success(value)
        },
      })
      return
    }
    case 'ResetLogin': {
      draft.status = Untriggered()
      return
    }
    case 'Logout': {
      localStorage.removeItem('token')
      sessionStorage.removeItem('token')
      draft.logoutStatus = Loading(null)
      draft.status = Untriggered()
      return
    }
    case 'LoggedOut': {
      either.match(action.result)({
        Left: ({ value }) => {
          draft.logoutStatus = Failure(value)
        },
        Right: ({ value }) => {
          draft.logoutStatus = Success(value)
        },
      })

      return
    }
    case 'NotifyForgotPassword': {
      draft.notifyForgotPasswordStatus = Loading(null)
      return
    }
    case 'ForgotPasswordNotified': {
      either.match(action.result)({
        Right: ({}) => {
          draft.notifyForgotPasswordStatus = Success(null)
        },
        Left: ({ value }) => {
          draft.notifyForgotPasswordStatus = Failure(value)
        },
      })
      return
    }
  }
}, initialState)
