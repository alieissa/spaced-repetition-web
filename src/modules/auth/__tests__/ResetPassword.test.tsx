/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { renderWithProviders } from 'src/utils/test-utils'
import ResetPassword from '../ResetPassword'
import { NAuth } from '../auth.types'

const resetPasswordUrl = `${process.env.REACT_APP_API_ENDPOINT}/users/reset-password`

const email = faker.internet.email({ firstName: 'Johnny', lastName: 'Bravo' })
export const handlers = [
  rest.post(resetPasswordUrl, (__, res, ctx) => {
    return res(ctx.json({ email }))
  }),
]

const setupForm = async ({
  password,
  confirmedPassword,
}: NAuth.ResetPasswordForm) => {
  const user = userEvent.setup()

  const passwordInputElement = screen.getByTestId(
    'reset-password-form-password-input',
  )
  await act(() => user.type(passwordInputElement, password))

  const confirmedPasswordInputElement = screen.getByTestId(
    'reset-password-form-confirmed-password-input',
  )
  await act(() => user.type(confirmedPasswordInputElement, confirmedPassword))
}

const submitForm = async () => {
  const resetPasswordBtn = screen.getByTestId('reset-password-form-btn')
  await act(() => resetPasswordBtn.click())
}

const expectElementToBeInDocument = async (testId: string) => {
  const element = await screen.findByTestId(testId)
  await waitFor(() => expect(element).toBeInTheDocument())
}
const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('ResetPassword', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<ResetPassword />)
      expect(asFragment()).toMatchSnapshot()
    })
    it('should display initial view', async () => {
      renderWithProviders(<ResetPassword />)
      await waitFor(() => expectElementToBeInDocument('reset-password-form'))
    })

    // Loading view not displayed so no test necessary
    // it('should display loading view', async ...)

    it('should display error view', async () => {
      // Assemble
      server.use(
        rest.post(resetPasswordUrl, (__, res, ctx) => {
          return res(ctx.status(422))
        }),
      )
      renderWithProviders(<ResetPassword />)
      const password = faker.internet.password()
      await setupForm({ password: password, confirmedPassword: password })

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('reset-password-form-error'),
      )
    })

    it('should display success view', async () => {
      // Assemble
      renderWithProviders(<ResetPassword />)
      const password = faker.internet.password()
      await setupForm({ password, confirmedPassword: password })

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('reset-password-form-success'),
      )
    })
  })

  describe('interaction', () => {
    it('should display error when password is invalid', async () => {
      // Assemble
      renderWithProviders(<ResetPassword />)
      const password = faker.internet.password({ length: 4 })
      await setupForm({ password, confirmedPassword: password })

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('reset-password-form-password-error'),
      )
    })

    it.todo(
      'should display error when confirmed password does not match password',
    )
  })
})
