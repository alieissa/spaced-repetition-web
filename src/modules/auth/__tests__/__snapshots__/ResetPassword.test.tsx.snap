// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui container"
  >
    <div
      class="justify-space-between bordered p-1r mb-1r"
    >
      <h1
        class="ui header"
        style="margin-top: 0px; margin-bottom: 0px;"
      >
        Spaced Reps
      </h1>
    </div>
    <form
      class="ui form flex-column bordered p-1r"
      data-testid="reset-password-form"
    >
      <div
        class="field"
      >
        <label
          for="rest-password-password-input"
        >
          Password
        </label>
        <div
          class="ui input"
          style="opacity: 1;"
        >
          <input
            data-testid="reset-password-form-password-input"
            id="rest-password-password-input"
            placeholder="Password"
            style="border-radius: unset;"
            title="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          for="reset-password-confirm-password-input"
        >
          Confirm password
        </label>
        <div
          class="ui input"
          style="opacity: 1;"
        >
          <input
            data-testid="reset-password-form-confirmed-password-input"
            id="reset-password-confirm-password-input"
            placeholder="Confirm password"
            style="border-radius: unset;"
            title="confirm password"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="ui button align-self-end"
        data-testid="reset-password-form-btn"
        style="border-radius: 0;"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
